src/game/stats.c

```diff
@@ -0,0 +1,328 @@
+                                    m_KillableItems[idx] = true;
+                                    m_LevelPickups += PIERRE_ITEMS;
+                                    m_LevelKillables += 1;
+                                }
+
+                                // Add killable if object triggered
+                                if (m_IfKillable[item->object_number]
+                                    && !m_KillableItems[idx]) {
+                                    m_KillableItems[idx] = true;
+                                    m_LevelKillables += 1;
+
+                                    // Add mercenary pickups
+                                    if (item->object_number == O_SKATEKID) {
+                                        m_LevelPickups += SKATEKID_ITEMS;
+                                    }
+                                    if (item->object_number == O_COWBOY) {
+                                        m_LevelPickups += COWBOY_ITEMS;
+                                    }
+                                    if (item->object_number == O_BALDY) {
+                                        m_LevelPickups += BALDY_ITEMS;
+                                    }
+                                }
+                            }
+                        } while (!(trigger & END_BIT));
+                        break;
+                    }
+                } while (!(type & END_BIT));
+            }
+        }
+    }
+}
+
+void Stats_ObserveRoomsLoad()
+{
+    m_CachedFloorArray =
+        GameBuf_Alloc(g_RoomCount * sizeof(FLOOR_INFO *), GBUF_ROOM_FLOOR);
+    for (int i = 0; i < g_RoomCount; i++) {
+        const ROOM_INFO *current_room_info = &g_RoomInfo[i];
+        int count = current_room_info->y_size * current_room_info->x_size;
+        m_CachedFloorArray[i] =
+            GameBuf_Alloc(count * sizeof(FLOOR_INFO), GBUF_ROOM_FLOOR);
+        memcpy(
+            m_CachedFloorArray[i], current_room_info->floor,
+            count * sizeof(FLOOR_INFO));
+    }
+}
+
+void Stats_ObserveItemsLoad()
+{
+    m_CachedItemCount = g_LevelItemCount;
+}
+
+void Stats_Init()
+{
+    for (int i = 0; m_KillableObjs[i] != NO_ITEM; i++) {
+        m_IfKillable[m_KillableObjs[i]] = true;
+    }
+}
+
+void Stats_CalculateStats()
+{
+    // Clear old values
+    m_LevelPickups = 0;
+    m_LevelKillables = 0;
+    m_LevelSecrets = 0;
+    memset(&m_KillableItems, 0, sizeof(m_KillableItems));
+
+    if (m_CachedItemCount) {
+        if (m_CachedItemCount > MAX_ITEMS) {
+            Shell_ExitSystem(
+                "Stats_CalculateStats(): Too Many g_Items being Loaded!!");
+            return;
+        }
+
+        for (int i = 0; i < m_CachedItemCount; i++) {
+            ITEM_INFO *item = &g_Items[i];
+
+            if (item->object_number < 0 || item->object_number >= O_NUMBER_OF) {
+                Shell_ExitSystemFmt(
+                    "Stats_CalculateStats(): Bad Object number (%d) on Item %d",
+                    item->object_number, i);
```

Comment:
Can we make it a non-fatal error and `LOG_ERROR` instead?

Note that `LOG_*` already includes function name, so we only need to directly log the error itself.